core/src/main/scala/cats/free/FreeSelective.scala

```diff
@@ -0,0 +1,146 @@
+            }
+        }
+
+      override def pure[A](x: A): FreeSelective[F, A] = Pure[F, A](x)
+
+      /**
+        * The canonical apS function agreement is enforced, since these are of
+        * necessity rigid selective functors, since the normal form that
+        * backs the free defintion only holds in the rigid case.
+        */
+      override def ap[A, B](ff: FreeSelective[F, A => B])(fa: FreeSelective[F, A]): FreeSelective[F, B] =
+        select(map[A, Either[A, B]](fa)(Left.apply[A, B]))(ff)
+    }
+
+    /**
+      * *
+      * *  From the original paper:
+      * *
+      * * instance Functor f => Selective (Select f) where
+      * *   select x (Pure y) = either y id <$> x -- Generalised identity
+      * *   select x (Select y z) = Select (select (f <$> x) (g <$> y)) (h <$> z) -- Associativity
+      * *     where
+      * *       f x = Right <$> x
+      * *       g y = \a -> bimap (,a) ($a) y
+      * *       h z = uncurry z
+      */
+    override def select[A, B](fab: FreeSelective[F, Either[A, B]])(fn: FreeSelective[F, A => B]): FreeSelective[F, B] =
+      fn match {
+        case fn: Pure[F, A => B] =>
+          map(fab)(_.fold(fn.run, identity))
+
+        case fn: Select.Aux[F, A, A => B] =>
+          val fx: FreeSelective[F, Either[A, Either[(A, A), B]]]  =
+            map(fab)(e => Right(e.leftMap(a => (a, a))))
+
+          val gy: FreeSelective[F, A => Either[(A, A), B]] =
+            map(fn.feab)(
+              (eit: Either[A, A => B]) =>
+                (a: A) =>
+                  Bifunctor[Either].bimap(eit)((_, a), f => f(a)))
+
+          val hz: F[((A, A)) => B] =
+            Functor[F].map(fn.fab)(
+              Function.uncurried[A, A, B] _
+                andThen Function.tupled[A, A, B])
+
+          new Select[F, B] {
+            type Source = (A, A)
+
+            val fab: F[Source => B] = hz
+            val feab: FreeSelective[F, Either[Source, B]] = select(fx)(gy)
+          }
+      }
+  }
+
+  /**
+    * -- Extract all possible effects from a selective computation
+    * getEffects :: Functor f => Select f a -> [f ()]
+    * getEffects = getOver . runSelect (Over . pure . void)
+    */
+  def liftSelect[F[_]: Functor, A](fa: F[A]): FreeSelective[F, A] =
+    new Select[F, A] {
+      type Source = Unit
+
+      val feab: FreeSelective[F, Either[Unit, A]] = Pure(Left())
+      val fab: F[Unit => A] = Functor[F].map(fa)(Function.const)
+    }
+
+  def runSelect[F[_], G[_]: Selective, A, B](nat: F ~> G)(free: FreeSelective[F, A]): G[A] =
+    free match {
+      case x: Pure[F, A] => Selective[G].pure(x.run)
+      case x: Select.Aux[F, A, B] =>
+        Selective[G].select(runSelect(nat)(x.feab))(nat(x.fab))
+    }
+
+  private def toOptionK[F[_]] = new FunctionK[F, Option] {
+    def apply[B](a: F[B]): Option[B] = None
+  }
+  def getPure[F[_], A](free: FreeSelective[F, A]): Option[A] =
+    runSelect[F, Option, A, A](toOptionK[F])(free)
+}
```

Comment:
We should probably also add `getEffects` but I haven't quite gotten around to doing that.